### DISCUSSION TITLE
dOnlyFans / Patreon

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#168] [Project proposals]
WIP

//cc @sa8

**Comment by aakoshh** [comment]
My initial ideas were along the following lines. 

The rootnet could be Filecoin where content producer store their digital assets with storage providers and pay them rent. The subnets are managing the payments and subscriptions between content producer and consumer. When a consumer pays a consumer, they should get some cryptographic key that allows them to go to the storage provider, pay them (via retrieval markets) and view what they subscribed to. It's not just payment; if I subscribe to a channel, I should get content specific keys encrypted to myself each time there's something new.

I thought Filecoin would be the root so checkpoints propagate upwards and allow proofs about subnets states to be included, such as proof of payment. But it would possibly be too slow for good UX to have to wait until the checkpoint travels all the way up. Maybe for subscription content, but if I buy I video I want to watch it now, not in a few hours.

One complication is that consumers might not want to make it public which content they are subscribed to. But we also don't want to become a privacy coin and be banned. To combat this, I thought about the following lines:
* There should be some trusted (e.g. by law enforcement) content auditor services that the content producers would pay to vet their uploads, to testify that they are not illegal like child pornography. The system should be able to provide a Zero Knowledge proof that can later be used to say that "as a user I paid for some content - not saying what - and here's a ZK proof showing that it had been audited and it's nothing illegal". See similar building blocks for finance in [Findora](https://findora.org/wp-content/uploads/2020/12/Findora_Litepaper_3.2_Final_Clean.pdf), such as "KYC for this undisclosed client exist".
* In the subnet, there could be an Escrow account to decouple consumer from producer. As a consumer I would send payment to the Escrow, including a Zero Knowledge proof that I am paying for a content that exist on Filecoin and it is legal. This is to ensure law enforcement can be convinced that even though they don't know who I'm paying exactly, it's okay. 
* Once payment becomes part of the subnet state, the consumer would produce a Zero Knowledge proof (or pay some external service to do it for me) that proves that there exist a payment to the Escrow with at the payment amount. They would encrypt this with the keys of the producer and put it on chain (but not including their public key or account ID), or send it to them directly. (This would reveal their identity to the producer, but I'm not sure what would be a way around that).
* The producer would find this ZK proof and a corresponding nullifier (as in the ZCash model) and show it to the Escrow, asking for the payment amount to be credited to their account. At the same time they would attach a cryptographic authorization key encrypted to the consumer, which they can take to Filecoin. 

So, the Escrow is effectively a Mixer, which as Guy pointed out with Blender can get banned for what it achieves. It's the ZK proof about the legality of the content that can save it. The reason it would still be required is so people can't get blackmailed for the content they watch.

**Reply by cryptoAtwill** [reply]
Interesting discussion, just found out about it. @aakoshh I do have a question about `One complication is that consumers might not want to make it public which content they are subscribed to`. Since wallet is so easy to get and assuming on chain identity is completely insulated from off chain community, then a wallet should not reveal the offchain identity of the user. Wouldn't a wallet be private enough since it cannot be traced back to the offchain user?

Also, I think this is a good technical use case for scaling distributed cdn and perhaps also the access control on top of cdn.

**Reply by sa8** [reply]
There is a large body of literature on de-anonymization techniques on blockchains so I do not think it is a safe assumption that "on chain identity is completely insulated from off chain community."
